client.go

```diff
@@ -958,3 +937,100 @@ func (c *TunaSessionClient) ServicePort() (int, error) {
 	}
 	return 0, errors.New("no tuna exit available")
 }
+
+func (c *TunaSessionClient) dialAConn(ctx context.Context, remoteAddr string, sessionID []byte, i int, addr *PubAddr, config *nkn.DialConfig) (conn *Conn, err error) {
+	if addr == nil || len(addr.IP) == 0 || addr.Port == 0 {
+		return nil, fmt.Errorf("dialAConn wrong params addr")
+	}
+
+	dialer := &net.Dialer{}
+	netConn, err := dialer.DialContext(ctx, "tcp", fmt.Sprintf("%s:%d", addr.IP, addr.Port))
+	if err != nil {
+		return nil, err
+	}
+	conn = newConn(netConn)
+
+	err = writeMessage(conn, []byte(c.addr.String()), time.Duration(config.DialTimeout)*time.Millisecond)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	metadata := &pb.SessionMetadata{
+		Id: sessionID,
+	}
+	metadataRaw, err := proto.Marshal(metadata)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	buf, err := c.encode(metadataRaw, remoteAddr)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	err = writeMessage(conn, buf, time.Duration(config.DialTimeout)*time.Millisecond)
+	if err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	sessKey := sessionKey(remoteAddr, sessionID)
+	c.Lock()
+	conns, ok := c.sessionConns[sessKey]
+	if !ok {
+		conns = make(map[string]*Conn)
+	}
+	conns[connID(i)] = conn
+	c.sessionConns[sessKey] = conns
+	c.Unlock()
+
+	return conn, nil
+}
+
+func (c *TunaSessionClient) reconnect(remoteAddr string, sessionID []byte, i int, pubAddrs *PubAddrs, config *nkn.DialConfig) (conn *Conn, err error) {
+	if c.IsClosed() {
+		return nil, ErrClosed
+	}
+	sessKey := sessionKey(remoteAddr, sessionID)
+	c.RLock()
+	sess := c.sessions[sessKey]
+	c.RUnlock()
+	if sess == nil || sess.IsClosed() {
+		return nil, ncp.ErrSessionClosed
+	}
+
+	ctx := context.Background()
+	var cancel context.CancelFunc
+	if config.DialTimeout > 0 {
+		ctx, cancel = context.WithTimeout(ctx, time.Duration(config.DialTimeout)*time.Millisecond)
+		defer cancel()
+	}
+
+	newPubAddrs, err := c.getPubAddrsFromRemote(ctx, remoteAddr)
+	if err != nil {
+		return nil, err
+	}
+	if len(newPubAddrs.Addrs) > i {
+		conn, err = c.dialAConn(ctx, remoteAddr, sessionID, i, newPubAddrs.Addrs[i], config)
+		if err == nil {
+			return conn, nil
+		}
+	}
+	return nil, err
+}
+
+// For testing only, after PR is approved, we can delete it.
```

Comment:
We can keep it because test code calls this function

Reply:
Yes, comment is removed.